### PR TITLE
Add support for building on circleci

### PR DESCRIPTION
**Pull request by springmeyer**
This adds a circle.yml and script to build osrm-backend on ubuntu with clang-3.8.

This is an experiment to enable us to ask questions like:
- How fast do builds run on circleci vs travis?
- Will we hit the circleci limits of the free account?
- How much might it cost to enable circle CI OS X builds?

Next actions:
- [ ] @TheMarex: Enable circleci to see if this passes (it should and if not, it should be very close to working as I verified a passing build under my personal account: https://circleci.com/gh/springmeyer/osrm-backend/11)
- [ ] After working, then we could try to get OS X working (should require very minimal changes)

### CODE REVIEW

File: circle.yml [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+    - "~/.ccache"
+    - "~/.apt-cache"
+    - "mason_packages"
+  pre:
+    # https://discuss.circleci.com/t/add-ability-to-cache-apt-get-programs/598/3
+    - sudo rm -rf /var/cache/apt/archives && sudo ln -s ~/.apt-cache /var/cache/apt/archives && mkdir -p ~/.apt-cache/partial
+    - sudo add-apt-repository -y ppa:ubuntu-toolchain-r/test
+    - sudo apt-get update -y
+  override:
+    - sudo apt-get install -y libstdc++6 libstdc++-5-dev
```
Comment (daniel-j-h) [inline circle.yml]: What is the reason we need both, and only the `-dev` version for 5?

File: scripts/circleci.sh [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+      -DCMAKE_LIBRARY_PATH=${MASON_HOME}/lib \
+      -DCMAKE_BUILD_TYPE=${BUILD_TYPE} \
+      -DBoost_USE_STATIC_LIBS=ON \
+      -DBUILD_TOOLS=1 \
+      -DENABLE_CCACHE=ON \
+      -DBUILD_SHARED_LIBS=${BUILD_SHARED_LIBS:-OFF} \
+      -DCOVERAGE=${COVERAGE:-OFF} \
+      ${CMAKE_EXTRA_ARGS}
+
+    make osrm-extract --jobs=3
```
Comment (daniel-j-h) [inline scripts/circleci.sh]: Is this a copy&paste leftover? The line below already builds all binary targets, and uses the `JOBS` variable.

**Comment by daniel-j-h** [comment]
@springmeyer what is the status of this pr?

**Comment by springmeyer** [comment]
@daniel-j-h - just rebased and updated. Status is that it should be ready to test by enabling circleci. Because enabling circleci will enable builds for all branches, we should merge this as is and continue to work on it in master once circleci is enabled (by @TheMarex).

**Comment by TheMarex** [comment]
So can we use CircleCI only for OSX without making our linux builds fail?

**Comment by springmeyer** [comment]
> So can we use CircleCI only for OSX without making our linux builds fail? Just re-enabled builds.

@TheMarex yes, by merging this PR and ensuring that "On" is checked for `Build OS X Project` at https://circleci.com/gh/Project-OSRM/osrm-backend/edit#build-environment

**Comment by springmeyer** [comment]
per chat with @TheMarex my next action here is:

  - [ ] Split the mason setup from the build/test (this will allow future work to hook into the mason deps/setup)
  - [ ]  Ensure the build/test setup works on(Because we can):
     - circleci OS X
     - circleci linux
     - travis OS X
     - travis linux
  
/cc @kkaefer @daniel-j-h

**Comment by springmeyer** [comment]
Capturing further discussion with @TheMarex that:

  - I plan to refactor the .travis.yml code into standalone bash
  - This will then be able to be run on both circle, travis, and locally
  - We will carefully use `travis_fold` to ensure that the travis logs are cleanly organized and collapsable even after all logic is moved to bash (we'll learn from https://github.com/mapbox/mapbox-gl-native/blob/master/scripts/travis_helper.sh)

**Comment by TheMarex** [comment]
@springmeyer what is the current state here? What is left to do?

**Comment by springmeyer** [comment]
@TheMarex all the above bullets remain. I've only had time to fix 90cb57e52c72f4ce3fb17951dbf0a9656414f0e3, which was preventing some osx builds from working locally. Next week is full for me but I can get back to pushing this forward the following week.

**Comment by TheMarex** [comment]
Out build system has seen some bigger refactor since this PR has been opened. While we are still not too happy with Travis CI is works reasonably well.

Closing this as outdated.

**Comment by springmeyer** [comment]
👍

**Pull request closed without merging**